test/check-machines-hostdevs

```diff
@@ -85,6 +85,84 @@ class TestMachinesHostDevs(VirtualMachinesCase):
 
         b.wait_in_text("#vm-subVmTest1-hostdev-1-source #1-slot", "0000:00:0f.0")
 
+    def testHostDevAdd(self):
+        b = self.browser
+        m = self.machine
+
+        self.createVm("subVmTest1")
+
+        self.login_and_go("/machines")
+        b.wait_in_text("body", "Virtual machines")
+
+        polkit = """
+polkit.addRule(function(action, subject) {
+    if (subject.user == "admin") {
+        return polkit.Result.YES;
+    }
+});
+"""
+        m.execute("echo '{0}' >> /etc/polkit-1/rules.d/80-libvirt.rules".format(polkit))
+
+        self.goToVmPage("subVmTest1")
+        b.wait_visible("#vm-subVmTest1-hostdevs")
+
+        # Shut off domain
+        self.performAction("subVmTest1", "forceOff")
+
+        class HostDevAddDialog(object):
+            def __init__(
+                self, test_obj, dev_type="usb_device", dev_id=0, vm_dev_id=1
+            ):
+                self.test_obj = test_obj
+                self.dev_type = dev_type
+                self.dev_id = dev_id
+                self.vm_dev_id = vm_dev_id
+                self._vendor = None
+                self._model = None
+
+            def execute(self):
+                self.open()
+                self.fill()
+                self.add()
+                self.verify()
+
+            def open(self):
+                b.wait_not_present("#vm-subVmTest1-hostdev-{0}-product".format(self.vm_dev_id))
+                b.click("button#vm-subVmTest1-hostdevs-add")
+                b.wait_in_text(".pf-c-modal-box .pf-c-modal-box__header .pf-c-modal-box__title", "Add host device to VM")
+
+            def fill(self):
+                b.click("input#{0}".format(self.dev_type))
+                b.set_checked(".pf-c-table input[name='checkrow{0}']".format(self.dev_id), True)
+                self._model = b.text("#vm-subVmTest1-hostdevs-dialog table tbody tr:nth-child({0}) td:nth-child(2)".format(self.dev_id + 1))
+                self._vendor = b.text("#vm-subVmTest1-hostdevs-dialog table tbody tr:nth-child({0}) td:nth-child(3)".format(self.dev_id + 1))
+
+            def cancel(self):
+                b.click(".pf-c-modal-box__footer button:contains(Cancel)")
+                b.wait_not_present("#vm-subVmTest1-hostdevs-dialog")
+
+            def add(self):
+                b.click(".pf-c-modal-box__footer button:contains(Add)")
+                b.wait_not_present("#vm-subVmTest1-hostdevs-dialog")
+
+            def verify(self):
+                b.wait_visible("#vm-subVmTest1-hostdev-{0}-product".format(self.vm_dev_id))
+                b.wait_in_text("#vm-subVmTest1-hostdev-{0}-product".format(self.vm_dev_id), self._model)
+                b.wait_in_text("#vm-subVmTest1-hostdev-{0}-vendor".format(self.vm_dev_id), self._vendor)
```

Comment:
I know it's hard to detect that the devices are really visible inside the VM. But please at least make sure that (1) the dumped XML now contains the device, and (2) the VM still boots (to guard against syntax/XML structure errors and such)

Reply:
I think this is resolved now -- verify_backend() does that.